target/linux/ipq806x/patches-5.10/097-2-drivers-net-dsa-qca8k-fix-wrong-mtu-configuration.patch

```diff
@@ -0,0 +1,57 @@
+From 8caff9db809d96af154bdda473d7382ebe7d8ef7 Mon Sep 17 00:00:00 2001
+From: Ansuel Smith <ansuelsmth@gmail.com>
+Date: Sun, 7 Mar 2021 18:06:26 +0100
+Subject: [PATCH] drivers : net: dsa: qca8k: fix wrong mtu configuration
+
+qca8k switch requires additional overhead for the double vlan and
+atheros header used to tag dsa packet. Fix this and mimic the original
+ar8337 switch driver.
+
+Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>
+---
+ drivers/net/dsa/qca8k.c | 8 ++++++--
+ drivers/net/dsa/qca8k.h | 2 ++
+ 2 files changed, 8 insertions(+), 2 deletions(-)
+
+diff --git a/drivers/net/dsa/qca8k.c b/drivers/net/dsa/qca8k.c
+index a72e89019c91..765ca87c1182 100644
+--- a/drivers/net/dsa/qca8k.c
++++ b/drivers/net/dsa/qca8k.c
+@@ -859,7 +859,9 @@ qca8k_setup(struct dsa_switch *ds)
+ 	/* Setup our port MTUs to match power on defaults */
+ 	for (i = 0; i < QCA8K_NUM_PORTS; i++)
+ 		priv->port_mtu[i] = ETH_FRAME_LEN + ETH_FCS_LEN;
+-	qca8k_write(priv, QCA8K_MAX_FRAME_SIZE, ETH_FRAME_LEN + ETH_FCS_LEN);
++	qca8k_write(priv, QCA8K_MAX_FRAME_SIZE,
++		    ETH_FRAME_LEN + ETH_FCS_LEN + QCA8K_ATHEROS_HEADER +
++			    QCA8K_DOUBLE_VLAN);
```

Comment:
According to the datasheet

> The value is for normal packet, it is added 4 by MAC if
> support VLAN, added 8 for double VLAN, and added 2 for
> Atheros header.

Doesn't that mean VLAN and DSA tag overhead don't count towards the MAX_FRAME_SIZE?

Reply:
The datasheet is not so clear about this.
If they doesn't count then why the original ar8327 driver set the max_frame_size with them?
Any idea how to test this?

Reply:
Hi , 
I Built my own with your patches, my c2600 works great here but I share my builds with someone else, and he has the r7800 and it seems , one port does not work and some problems with banip and adblock.

Is it the first one to see that about adblock and ban ip?
Thanks

Reply:
> first one to see that about adblock and ban ip?

I think that adblock works quite normally on my R7800.  
(It is mainly about DNS functionality, and it is only lightly tied to the network interfaces.)

Both banip and adblock and sqm and ... may require some config change if a specific wan interface is set as trigger interface.

Reply:
> Any idea how to test this?

Just ping the router with `-s <payload length>` option (or `-l` on Windows)

cc @u1f35c

Reply:
I don't believe there's a need to change this from upstream; I'm running with the unmodified code on a network that has VLANs in operation and don't see any MTU problems:

`$ ping -c 3 -M do -s 1472 192.168.35.1
PING 192.168.35.1 (192.168.35.1) 1472(1500) bytes of data.
1480 bytes from 192.168.35.1: icmp_seq=1 ttl=64 time=0.689 ms
1480 bytes from 192.168.35.1: icmp_seq=2 ttl=64 time=0.686 ms
1480 bytes from 192.168.35.1: icmp_seq=3 ttl=64 time=0.672 ms

--- 192.168.35.1 ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 34ms
rtt min/avg/max/mdev = 0.672/0.682/0.689/0.022 ms
`

Reply:
so... the original code was wrong and the datasheet say that the overhead is added internally?

Reply:
Looks like it. Setting the switch MTU higher than required is unlikely to result in any noticeable issues, whereas if it was too low you'd see silently dropped packets, so my guess is the original driver got it wrong and as that didn't cause problems no one noticed.